<<<<<<< HEAD
# Disallows invocation of `require()` (no-require-imports)

Prefer the newer ES6-style imports over `require()`.

## Rule Details

Examples of **incorrect** code for this rule:

```ts
var lib = require('lib');
let lib2 = require('lib2');
var lib5 = require('lib5'),
  lib6 = require('lib6');
import lib8 = require('lib8');
```

Examples of **correct** code for this rule:

```ts
import { l } from 'lib';
var lib3 = load('not_an_import');
var lib4 = lib2.subImport;
var lib7 = 700;
import lib9 = lib2.anotherSubImport;
import lib10 from 'lib10';
=======
---
description: 'Disallow invocation of `require()`.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-require-imports** for documentation.

Prefer the newer ES6-style imports over `require()`.

## Examples

<!--tabs-->

### âŒ Incorrect

```ts
const lib1 = require('lib1');
const { lib2 } = require('lib2');
import lib3 = require('lib3');
```

### âœ… Correct

```ts
import * as lib1 from 'lib1';
import { lib2 } from 'lib2';
import * as lib3 from 'lib3';
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
```

## When Not To Use It

<<<<<<< HEAD
If you don't care about TypeScript module syntax, then you will not need this rule.

## Compatibility

- TSLint: [no-require-imports](https://palantir.github.io/tslint/rules/no-require-imports/)
=======
If you don't care about using newer module syntax, then you will not need this rule.

## Related To

- [`no-var-requires`](./no-var-requires.md)
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
