<<<<<<< HEAD
# Warns if a type assertion does not change the type of an expression (no-unnecessary-type-assertion)

This rule prohibits using a type assertion that does not change the type of an expression.

## Rule Details

This rule aims to prevent unnecessary type assertions.

Examples of **incorrect** code for this rule:
=======
---
description: 'Disallow type assertions that do not change the type of an expression.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-unnecessary-type-assertion** for documentation.

TypeScript can be told an expression is a different type than expected using `as` type assertions.
Leaving `as` assertions in the codebase increases visual clutter and harms code readability, so it's generally best practice to remove them if they don't change the type of an expression.
This rule reports when a type assertion does not change the type of an expression.

## Examples

<!--tabs-->

### âŒ Incorrect
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd

```ts
const foo = 3;
const bar = foo!;
```

```ts
const foo = <3>3;
```

```ts
type Foo = 3;
const foo = <Foo>3;
```

```ts
type Foo = 3;
const foo = 3 as Foo;
```

```ts
function foo(x: number): number {
  return x!; // unnecessary non-null
}
```

<<<<<<< HEAD
Examples of **correct** code for this rule:
=======
### âœ… Correct
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd

```ts
const foo = <number>3;
```

```ts
const foo = 3 as number;
```

```ts
<<<<<<< HEAD
=======
const foo = 'foo' as const;
```

```ts
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
function foo(x: number | undefined): number {
  return x!;
}
```

<<<<<<< HEAD
### Options

This rule optionally takes an object with a single property `typesToIgnore`, which can be set to a list of type names to ignore.

For example, with `@typescript-eslint/no-unnecessary-type-assertion: ["error", { typesToIgnore: ['Foo'] }]`, the following is **correct** code":
=======
## Options

### `typesToIgnore`

With `@typescript-eslint/no-unnecessary-type-assertion: ["error", { typesToIgnore: ['Foo'] }]`, the following is **correct** code":
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd

```ts
type Foo = 3;
const foo: Foo = 3;
```

## When Not To Use It

If you don't care about having no-op type assertions in your code, then you can turn off this rule.
<<<<<<< HEAD

## Related to

- TSLint: ['no-unnecessary-type-assertion`](https://palantir.github.io/tslint/rules/no-unnecessary-type-assertion/)
=======
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
