<<<<<<< HEAD
# Warns when a namespace qualifier is unnecessary. (no-unnecessary-qualifier)

## Rule Details

This rule aims to let users know when a namespace or enum qualifier is unnecessary,
whether used for a type or for a value.

Examples of **incorrect** code for this rule:

```ts
namespace A {
  export type B = number;
  const x: A.B = 3;
}
```

```ts
namespace A {
  export const x = 3;
  export const y = A.x;
}
```
=======
---
description: 'Disallow unnecessary namespace qualifiers.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-unnecessary-qualifier** for documentation.

Members of TypeScript enums and namespaces are generally retrieved as qualified property lookups: e.g. `Enum.member`.
However, when accessed within their parent enum or namespace, the qualifier is unnecessary: e.g. just `member` instead of `Enum.member`.
This rule reports when an enum or namespace qualifier is unnecessary.

## Examples

<!--tabs-->

### âŒ Incorrect
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd

```ts
enum A {
  B,
  C = A.B,
}
```

```ts
namespace A {
<<<<<<< HEAD
  export namespace B {
    export type T = number;
    const x: A.B.T = 3;
  }
}
```

Examples of **correct** code for this rule:

```ts
namespace X {
  export type T = number;
}

namespace Y {
  export const x: X.T = 3;
}
```

```ts
enum A {
  X,
  Y,
}

enum B {
  Z = A.X,
=======
  export type B = number;
  const x: A.B = 3;
}
```

### âœ… Correct

```ts
enum A {
  B,
  C = B,
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
}
```

```ts
<<<<<<< HEAD
namespace X {
  export type T = number;
  namespace Y {
    type T = string;
    const x: X.T = 0;
  }
=======
namespace A {
  export type B = number;
  const x: B = 3;
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
}
```

## When Not To Use It

<<<<<<< HEAD
If you don't care about having unneeded namespace or enum qualifiers, then you don't need to use this rule.

## Further Reading

- TSLint: [no-unnecessary-qualifier](https://palantir.github.io/tslint/rules/no-unnecessary-qualifier/)
=======
If you don't care about having unneeded enum or namespace qualifiers, then you don't need to use this rule.
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
