<<<<<<< HEAD
# Disallows the use of require statements except in import statements (no-var-requires)

In other words, the use of forms such as `var foo = require("foo")` are banned. Instead use ES6 style imports or `import foo = require("foo")` imports.

## Rule Details

Examples of **incorrect** code for this rule:
=======
---
description: 'Disallow `require` statements except in import statements.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-var-requires** for documentation.

In other words, the use of forms such as `var foo = require("foo")` are banned. Instead use ES6 style imports or `import foo = require("foo")` imports.

## Examples

<!--tabs-->

### âŒ Incorrect
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd

```ts
var foo = require('foo');
const foo = require('foo');
let foo = require('foo');
```

<<<<<<< HEAD
Examples of **correct** code for this rule:
=======
### âœ… Correct
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd

```ts
import foo = require('foo');
require('foo');
<<<<<<< HEAD
=======
import foo from 'foo';
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
```

## When Not To Use It

<<<<<<< HEAD
If you don't care about TypeScript module syntax, then you will not need this rule.

## Compatibility

- TSLint: [no-var-requires](https://palantir.github.io/tslint/rules/no-var-requires/)
=======
If you don't care about using newer module syntax, then you will not need this rule.

## Related To

- [`no-require-imports`](./no-require-imports.md)
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
