<<<<<<< HEAD
# Require the use of the `namespace` keyword instead of the `module` keyword to declare custom TypeScript modules. (prefer-namespace-keyword)

In an effort to prevent further confusion between custom TypeScript modules and the new ES2015 modules, starting
with TypeScript v1.5 the keyword `namespace` is now the preferred way to declare custom TypeScript modules.

## Rule Details

This rule aims to standardise the way modules are declared.
=======
---
description: 'Require using `namespace` keyword over `module` keyword to declare custom TypeScript modules.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/prefer-namespace-keyword** for documentation.

TypeScript historically allowed a form of code organization called "custom modules" (`module Example {}`), later renamed to "namespaces" (`namespace Example`).

Namespaces are an outdated way to organize TypeScript code.
ES2015 module syntax is now preferred (`import`/`export`).

For projects still using custom modules / namespaces, it's preferred to refer to them as namespaces.
This rule reports when the `module` keyword is used instead of `namespace`.

> This rule does not report on the use of TypeScript module declarations to describe external APIs (`declare module 'foo' {}`).

## Examples

<!--tabs-->

### ‚ùå Incorrect

```ts
module Example {}
```

### ‚úÖ Correct

```ts
namespace Example {}

declare module 'foo' {}
```

<!--/tabs-->
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd

## When Not To Use It

If you are using the ES2015 module syntax, then you will not need this rule.

## Further Reading

- [Modules](https://www.typescriptlang.org/docs/handbook/modules.html)
- [Namespaces](https://www.typescriptlang.org/docs/handbook/namespaces.html)
- [Namespaces and Modules](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html)
<<<<<<< HEAD

## Compatibility

- TSLint: [no-internal-module](https://palantir.github.io/tslint/rules/no-internal-module/)
=======
>>>>>>> 21d7be62d554915787485641c4ea1187492767dd
